src/main/java/captain/cybot/adventure/backend/service/UserServiceImpl.java

```diff
@@ -0,0 +1,119 @@
+import java.util.ArrayList;
+import java.util.Collection;
+
+@Service
+@AllArgsConstructor
+@Transactional
+@Slf4j
+public class UserServiceImpl implements UserService, UserDetailsService {
+
+    private UserRepository userRepository;
+    private RoleRepository roleRepository;
+    private CosmeticRepository cosmeticRepository;
+
+    @Autowired
+    public BCryptPasswordEncoder passwordEncoder() {
+        return new BCryptPasswordEncoder();
+    }
+
+    @Autowired
+    public PasswordConstraintValidator passwordConstraintValidator() {
+        return new PasswordConstraintValidator();
+    }
+
+    @Override
+    public User saveUser(User user) throws UserAlreadyExistsException, InvalidRoleException, PasswordInvalidException {
+        if(usernameAlreadyExists(user.getUsername())) {
+            throw new UserAlreadyExistsException("Username already exists.");
+        } else if(emailAlreadyExists(user.getEmail())) {
+            throw new UserAlreadyExistsException("Email already exists.");
+        } else if (!passwordConstraintValidator().isValid(user.getPassword())) {
+            throw new PasswordInvalidException("Password must contain the following: \n-8 to 24 characters" +
+                    "\n-ONE lowercase character \n-ONE uppercase character" +
+                    "\n-ONE digit character \n-ONE special character");
+        }
+        user.setPassword(passwordEncoder().encode(user.getPassword()));
+        User newUser = userRepository.save(user);
+        log.info(newUser.getUsername() + "'s account has been created.");
+        addRoleToUser(newUser.getUsername(), ROLES.ROLE_USER.toString());
+        setDefaultCosmetic(newUser.getUsername());
+        return newUser;
+    }
+
+    @Override
+    public User getUser(String username) {
+        log.info("Fetching user " + username);
+        return userRepository.findByUsername(username);
+    }
+
+    private boolean usernameAlreadyExists(String username) {
+        return userRepository.findByUsername(username) != null;
+    }
+
+    private boolean emailAlreadyExists(String email) {
+        return userRepository.findByEmail(email) != null;
+    }
+
+    @Override
+    public void addRoleToUser(String username, String roleName) throws UsernameNotFoundException, InvalidRoleException {
+        User user = userRepository.findByUsername(username);
+        Role role = roleRepository.findByName(roleName);
+        if (user != null && role != null) {
+            user.getRoles().add(role);
+            log.info(role.getName() + " has been added to " + user.getUsername());
+        } else if (user == null) {
+            throw new UsernameNotFoundException("User not found with username: " + username);
+        } else {
+            throw new InvalidRoleException("Invalid Role name: " + roleName +
+                    " Valid Role names: " + ROLES.ROLE_USER + " AND " + ROLES.ROLE_ADMIN);
+        }
+    }
+
+    public void setDefaultCosmetic(String username) {
```

Comment:
Should there be a setCosmetic(username, cosmetic) for non-default options?

Reply:
Yes, I am sure that this method will change or a new one will be added down the line when we start working on the worlds.